core/trino-main/src/main/java/io/trino/sql/planner/LocalDynamicFilterConsumer.java

```diff
@@ -13,105 +13,123 @@
     {
         this.buildChannels = requireNonNull(buildChannels, "buildChannels is null");
         this.filterBuildTypes = requireNonNull(filterBuildTypes, "filterBuildTypes is null");
         verify(buildChannels.keySet().equals(filterBuildTypes.keySet()), "filterBuildTypes and buildChannels must have same keys");
-
         requireNonNull(collectors, "collectors is null");
         checkArgument(!collectors.isEmpty(), "collectors is empty");
-        this.collectors = collectors;
-        this.partitions = new ArrayList<>();
+        this.collectors = ImmutableList.copyOf(collectors);
+        this.domainSizeLimitInBytes = requireNonNull(domainSizeLimit, "domainSizeLimit is null").toBytes();
     }
 
     @Override
-    public void addPartition(TupleDomain<DynamicFilterId> tupleDomain)
+    public void addPartition(TupleDomain<DynamicFilterId> domain)
     {
+        if (collected) {
+            return;
+        }
+
+        summaryDomains.add(domain);
+        unionSummaryDomains();
+
         TupleDomain<DynamicFilterId> result;
         synchronized (this) {
+            verify(expectedPartitionCount == null || collectedPartitionCount < expectedPartitionCount);
+
             if (collected) {
+                summaryDomains.clear();
                 return;
             }
-            requireNonNull(partitions, "partitions is null");
-            // Called concurrently by each DynamicFilterSourceOperator instance (when collection is over).
-            verify(expectedPartitionCount == PARTITION_COUNT_INITIAL_VALUE || partitions.size() < expectedPartitionCount);
-            // NOTE: may result in a bit more relaxed constraint if there are multiple columns and multiple rows.
-            // See the comment at TupleDomain::columnWiseUnion() for more details.
-            partitions.add(tupleDomain);
-            if (tupleDomain.isAll()) {
-                result = tupleDomain;
+            collectedPartitionCount++;
+
+            boolean allPartitionsCollected = expectedPartitionCount != null && collectedPartitionCount == expectedPartitionCount;
+            if (allPartitionsCollected) {
+                // run final compaction as previous concurrent compactions may have left more than a single domain
+                unionSummaryDomains();
             }
-            else if (partitions.size() == expectedPartitionCount) {
-                // No more partitions are left to be processed.
-                if (partitions.isEmpty()) {
-                    result = TupleDomain.none();
+
+            boolean sizeLimitExceeded = false;
+            TupleDomain<DynamicFilterId> summary = summaryDomains.poll();
+            // summary can be null as another concurrent summary compaction may be running
+            if (summary != null) {
+                if (summary.getRetainedSizeInBytes(DynamicFilterId::getRetainedSizeInBytes) > domainSizeLimitInBytes) {
+                    summary = summary.simplify();
```

Comment:
Think we need to `simplify(1)` to ensure that this always does some compaction if possible

Reply:
`simplify(1)` is pretty aggressive (will compact as soon as there's only a single value). Should we try incremental compactions?

Reply:
simplify will either result in a min-max range or do nothing. Since this is a last ditch attempt to get a useable filter, shouldn't we try to ensure that the call to `simplify` isn't a no-op ?